packages/angular/src/wallet/wallet.store.ts

```diff
@@ -0,0 +1,295 @@
+    readonly disconnecting$ = this.select((state) => state.disconnecting);
+    readonly wallet$ = this.select((state) => state.wallet);
+    readonly adapter$ = this.select((state) => state.adapter);
+    readonly publicKey$ = this.select((state) => state.publicKey);
+    readonly ready$ = this.select((state) => state.ready);
+    readonly anchorWallet$ = this.select(
+        this.publicKey$.pipe(isNotNull),
+        this.adapter$.pipe(isNotNull),
+        (publicKey, adapter) => {
+            return publicKey && 'signTransaction' in adapter && 'signAllTransactions' in adapter
+                ? {
+                      publicKey,
+                      signTransaction: (transaction: Transaction) => this.signTransaction(transaction)?.toPromise(),
+                      signAllTransactions: (transactions: Transaction[]) =>
+                          this.signAllTransactions(transactions)?.toPromise(),
+                  }
+                : undefined;
+        }
+    );
+
+    constructor(
+        @Optional()
+        @Inject(WALLET_CONFIG)
+        private _config: WalletConfig
+    ) {
+        super({
+            wallets: _config.wallets,
+            selectedWallet: null,
+            wallet: null,
+            adapter: null,
+            connected: false,
+            connecting: false,
+            disconnecting: false,
+            ready: false,
+            publicKey: null,
+            autoApprove: false,
+        });
+
+        if (!this._config) {
+            this._config = WALLET_DEFAULT_CONFIG;
+        } else {
+            this._config = {
+                ...WALLET_DEFAULT_CONFIG,
+                ...this._config,
+            };
+        }
+
+        const walletName = localStorage.getItem(this._localStorageKey);
+        const wallet = this._config.wallets.find(({ name }) => name === walletName);
+
+        if (wallet) {
+            this.selectWallet(walletName as WalletName);
+        } else if (this._config.wallets.length > 0) {
+            this.selectWallet(this._config.wallets[0].name);
+        }
+    }
+
+    readonly autoConnect = this.effect(() => {
+        return combineLatest([this.adapter$, this.ready$]).pipe(
+            filter(([adapter, ready]) => this._autoConnect && !!adapter && ready),
+            observeOn(asyncScheduler),
+            tap(() => this.connect())
+        );
+    });
+
+    readonly connect = this.effect((action$: Observable<void>) => {
+        return action$.pipe(
+            concatMap(() => of(null).pipe(withLatestFrom(this.state$, (_, state) => state))),
+            filter(({ connected, connecting, disconnecting }) => !connected && !connecting && !disconnecting),
+            tap(() => this.patchState({ connecting: true })),
+            concatMap(({ adapter, wallet, ready }) => {
+                if (!wallet || !adapter) {
+                    this.logError(new WalletNotSelectedError());
+                    return of(null);
```

Comment:
The reality is that throwing the error there was killing the entire stream and calling connect afterward wasn't working, so I decided to hack it that way, I can give it a second try.

Reply:
Hmm, okay. Yeah, I guess I just don't understand the expected behavior for these methods. In the React code they are really just vanilla JS, so they throw errors if there are errors.

Reply:
I don't like that hack, it gives me the creeps lol. I mean the method works, I have that working in a dapp but I don't like the fact that it's still a hack. 

Let me see if I can crack it down.

Reply:
My main issue is that if you return null, the code that wraps this doesn't know how to handle it.

For example, in the React code, you can do:
```ts
const { connect } = useWallet();
//...
try {
    await connect();
} catch (error) {
    if (error instanceof WalletNotSelectedError) {
        // do something specific
    }
}
```

This general pattern derives from web3-react: https://github.com/NoahZinsmeister/web3-react/blob/f5c91a368a7cc15d0ae801288212c384449eb464/packages/injected-connector/src/index.ts#L11-L17

Each wallet can throw errors with different levels of specificity and apps can handle them differently.

Reply:
This relates to [this comment](https://github.com/solana-labs/wallet-adapter/pull/42#discussion_r706651931), you can't actually call this.walletStore.connect() and *wait* for an error/response, that's handled by the component-store library. When you call `connect()`, you're just triggering the effect, then the store tries to connect, if it does connect it updates the state if not it handles the error internally, it's common to add the error to an internal state so consumers can "listen" to an errors stream, that sends every error.

Reply:
Ah, I see. 

For example, one common user flow in EVM apps is to try to connect, and if the user has Metamask installed but they don't have it configured to use the chain your app is on (say, BSC), you can actually catch the error, and then send a request to Metamask to use BSC instead of the chain they are on. The user accepts, they never see an error, the app state just shows them as connecting the whole time, the connection goes through, and the whole thing is basically seamless.

As long as there's an idiomatic way to handle the error like you showed in https://github.com/solana-labs/wallet-adapter/pull/42#discussion_r706657394 I think that makes sense then. It definitely weirds me out that you can't handle the error locally though.